File: 14.longest-common-prefix.js
/*
 * @lc app=leetcode id=14 lang=javascript
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string}
 */
function longestCommonPrefix(strs) {
  if (strs.length === 0) {
    return "";
  }

  const firstStr = strs[0];
  if (firstStr.length === 0) {
    return "";
  }

  let prefix;
  if (matchChar(strs, 0, firstStr[0])) {
    prefix = firstStr[0];
  } else {
    return "";
  }

  for (let i = 1; i < firstStr.length; i++) {
    const char = firstStr[i];
    if (matchChar(strs, i, char)) {
      prefix += char;
    }
  }

  return prefix;
}

function matchChar(strs, index, char) {
  for (let i = 1; i < strs.length; i++) {
    if (char !== strs[i][index]) {
      return false;
    }
  }

  return true;
}
// @lc code=end

File: 14.longest-common-prefix.rs
/*
 * @lc app=leetcode id=14 lang=rust
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        if strs.is_empty() {
            return String::from("");
        }

        let mut prefix = String::from("");
        let mut i = 0;
        loop {
            let mut c = None;
            for word in &strs {
                if i == word.len() {
                    return prefix;
                }

                match c {
                    None => c = word.chars().nth(i),
                    Some(c) => match word.chars().nth(i) {
                        Some(c2) if c == c2 => continue,
                        _ => return prefix,
                    },
                }
            }
            if let Some(ch) = c {
                prefix.push(ch);
            }

            i += 1;
        }
    }
}
// @lc code=end
